src/main/java/org/verdictdb/jdbc41/VerdictStreamResultSet.java

```diff
@@ -0,0 +1,1345 @@
+    //    }
+    //    else if (expected.equals("clob")) {
+    //      return actual.equals("clob");
+    //    }
+    //    else if (expected.equals("array")) {
+    //      return actual.equals("array");
+    //    }
+    //    else if (expected.equals("ref")) {
+    //      return actual.equals("ref");
+    //    }
+    //    else if (expected.equals("sqlxml")) {
+    //      return actual.equals("xml");
+    //    }
+    //    else if (expected.equals("rowid")) {
+    //      return actual.equals("rowid");
+    //    }
+    //    else if (expected.equals("nclob")) {
+    //      return actual.equals("nclob");
+    //    }
+    //    else return false;
+  }
+
+  @Override
+  public boolean isWrapperFor(Class<?> iface) throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public boolean last() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public void moveToCurrentRow() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  @Override
+  public void moveToInsertRow() throws SQLException {
+    throw new SQLFeatureNotSupportedException();
+  }
+
+  /**
+   * Require a mutex here. The stream is executed on the other thread. When the execution of the stream is finished,
+   * it will try to synchronize hasReadAllQueryResults flag by calling synchronized(hasReadAllQueryResults).
+   * And here are two cases: 1) next() executes first 2) the other thread executes first
+   *
+   * 1) If next() first executes, the flag hasReadAllQueryResults is still false. next() will try to take new queryResult
+   * from queryResults. As soon as, next doesn't acquire hasReadAllQueryResults flag. Then the other thread will take the
+   * acquire of hasReadAllQueryResults, append the queryresult and set the flag to be true. At that time, if another next() is called,
+   * it will get blocked at the line 'else if (queryResults.peek() == null && hasReadAllQueryResults)' until the other thread finishes
+   * its job.
+   *
+   * 2) If the other thread executes first, it will append the query result and set the flag to be true. Then waiting
+   * next() starts to execute. Since there are still queryResult in queryResults, next() will take it from the list so that
+   * the last queryResult won't be neglected. The second next() will return false, since queryResults.peek()==null and
+   * hasReadAllQueryResults==true.
+   *
+   * @return
+   * @throws SQLException
+   */
+  @Override
+  public boolean next() throws SQLException {
+    try {
+      // on the first call
+      if (queryResult == null && (!hasReadAllQueryResults || !queryResults.isEmpty())) {
+        queryResult = queryResults.take();
+        lastQueryResultIndex++;
+      }
+      boolean hasMore = queryResult.next();
+      if (hasMore) {
+        rowIndex++;
+        return true;
+      } else if (queryResults.peek() == null && hasReadAllQueryResults) {
```

Comment:
Is there any difference between hasReadAllQueryResults and isCompleted? (I know I suggested to change the name of the latter variable.)

Reply:
No, isn't isCompleted the same with hasReadAllQueryResults?

Reply:
Oh. I see. This line previously used isCompleted.